File: src/serialize.ts
/**
 * 结构化输出序列化：JSON / CSV。
 * 字段与终端表格列一一对应（驼峰命名）；值为原始数值（cacheRate 为 0-1 小数）。
 */
import { emptyTotals, type Totals, type SessionRow, type RequestRow } from "./aggregate.ts";
import type { WindowName } from "./cli.ts";

/** 序列化为 JSON 文本（totals 为单对象；sessions/requests 为 { window, rows }） */
export function serializeJson(
  window: WindowName,
  totals: Totals,
  sessions: SessionRow[],
  requests: RequestRow[],
): string {
  let body: unknown;
  if (window === "totals") {
    body = { window, ...totalsToObject(totals) };
  } else if (window === "sessions") {
    body = { window, rows: sessions.map(sessionToObject) };
  } else {
    body = { window, rows: requests.map(requestToObject) };
  }
  return JSON.stringify(body, null, 2) + "\n";
}

/** 序列化为 CSV 文本（首行表头，字段与 JSON 一致；空数据也输出表头） */
export function serializeCsv(
  window: WindowName,
  totals: Totals,
  sessions: SessionRow[],
  requests: RequestRow[],
): string {
  const rows: Record<string, unknown>[] =
    window === "totals"
      ? [{ window, ...totalsToObject(totals) }]
      : window === "sessions"
        ? sessions.map(sessionToObject)
        : requests.map(requestToObject);
  // 空数据：用窗口的字段结构推导表头，保证脚本可消费
  const headers = Object.keys(rows[0] ?? emptyRowFor(window));
  const lines = [
    headers.map(csvEscape).join(","),
    ...rows.map((r) => headers.map((h) => csvEscape(String(r[h]))).join(",")),
  ];
  return lines.join("\n") + "\n";
}

/** 窗口为空时用于推导表头的占位行 */
function emptyRowFor(window: string): Record<string, unknown> {
  if (window === "sessions") {
    return { sessionId: "", timestamp: "", cwd: "", model: "", ...totalsToObject(emptyTotals()) };
  }
  if (window === "requests") {
    return { sessionId: "", timestamp: "", model: "", ...totalsToObject(emptyTotals()) };
  }
  return totalsToObject(emptyTotals());
}

/** CSV 字段转义：含逗号/引号/换行的字段加引号包裹，内部引号双写 */
function csvEscape(s: string): string {
  return /[",\n\r]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

export function totalsToObject(t: Totals): Record<string, unknown> {
  return {
    requests: t.requests,
    input: t.input,
    output: t.output,
    cacheRead: t.cacheRead,
    cacheWrite: t.cacheWrite,
    reasoning: t.reasoning,
    totalTokens: t.totalTokens,
    cost: t.cost,
    cacheRate: t.cacheRate,
  };
}

export function sessionToObject(r: SessionRow): Record<string, unknown> {
  return {
    sessionId: r.sessionId,
    timestamp: r.timestamp,
    cwd: r.cwd,
    model: r.model,
    ...totalsToObject(r),
  };
}

export function requestToObject(r: RequestRow): Record<string, unknown> {
  return {
    sessionId: r.sessionId,
    timestamp: r.timestamp,
    model: r.model,
    ...totalsToObject(r),
  };
}

File: internal/serialize/serialize.go
package serialize

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/heihei0299/pi-session-anylize/internal/domain"
)

func SerializeJSON(v any) ([]byte, error) {
	return json.MarshalIndent(v, "", "  ")
}

func totalsToCSVFields(tot domain.Totals, includeStatus bool) []string {
	fields := []string{
		strconv.Itoa(tot.Requests),
		fmt.Sprintf("%.0f", tot.Input),
		fmt.Sprintf("%.0f", tot.Output),
		fmt.Sprintf("%.0f", tot.CacheRead),
		fmt.Sprintf("%.0f", tot.CacheWrite),
		fmt.Sprintf("%.0f", tot.Reasoning),
		fmt.Sprintf("%.0f", tot.TotalTokens),
		fmt.Sprintf("%.4f", tot.Cost),
		fmt.Sprintf("%.4f", tot.CacheRate),
	}
	if includeStatus {
		fields = append(fields, tot.CostStatus)
	}
	return fields
}

var metricCSVHeaders = []string{
	"requests", "input", "output", "cacheRead", "cacheWrite", "reasoning", "totalTokens", "cost", "cacheRate",
}

func metricHeaders(includeStatus bool) []string {
	headers := append([]string(nil), metricCSVHeaders...)
	if includeStatus {
		headers = append(headers, "costStatus")
	}
	return headers
}

func SerializeCSV(window string, data any) ([]byte, error) {
	buf := &bytes.Buffer{}
	w := csv.NewWriter(buf)

	switch window {
	case "totals":
		tot, ok := data.(domain.Totals)
		if !ok {
			if totPtr, ok2 := data.(*domain.Totals); ok2 {
				tot = *totPtr
			}
		}
		status := tot.CostStatus != ""
		_ = w.Write(append([]string{"window"}, metricHeaders(status)...))
		_ = w.Write(append([]string{"totals"}, totalsToCSVFields(tot, status)...))

	case "sessions":
		rows, _ := data.([]domain.SessionRow)
		status := anySessionStatus(rows)
		_ = w.Write(append([]string{"sessionId", "timestamp", "cwd", "model"}, metricHeaders(status)...))
		for _, r := range rows {
			_ = w.Write(append([]string{r.SessionId, r.Timestamp, r.Cwd, r.Model}, totalsToCSVFields(r.Totals, status)...))
		}

	case "requests":
		rows, _ := data.([]domain.RequestRow)
		status := anyRequestStatus(rows)
		_ = w.Write(append([]string{"sessionId", "timestamp", "model"}, metricHeaders(status)...))
		for _, r := range rows {
			_ = w.Write(append([]string{r.SessionId, r.Timestamp, r.Model}, totalsToCSVFields(r.Totals, status)...))
		}

	case "groups":
		rows, _ := data.([]domain.GroupRow)
		status := anyGroupStatus(rows)
		_ = w.Write(append([]string{"model", "cwd"}, metricHeaders(status)...))
		for _, r := range rows {
			_ = w.Write(append([]string{r.Model, r.Cwd}, totalsToCSVFields(r.Totals, status)...))
		}

	case "period":
		rows, _ := data.([]domain.PeriodRow)
		status := anyPeriodStatus(rows)
		_ = w.Write(append([]string{"period"}, metricHeaders(status)...))
		for _, r := range rows {
			_ = w.Write(append([]string{r.Period}, totalsToCSVFields(r.Totals, status)...))
		}
	}

	w.Flush()
	return buf.Bytes(), w.Error()
}

func anySessionStatus(rows []domain.SessionRow) bool {
	for _, row := range rows {
		if row.CostStatus != "" {
			return true
		}
	}
	return false
}

func anyRequestStatus(rows []domain.RequestRow) bool {
	for _, row := range rows {
		if row.CostStatus != "" {
			return true
		}
	}
	return false
}

func anyGroupStatus(rows []domain.GroupRow) bool {
	for _, row := range rows {
		if row.CostStatus != "" {
			return true
		}
	}
	return false
}

func anyPeriodStatus(rows []domain.PeriodRow) bool {
	for _, row := range rows {
		if row.CostStatus != "" {
			return true
		}
	}
	return false
}
